### PR TITLE
add straight line functionality

### LINKED ISSUE #578: [Feature request]: Add shape detection for free hand
### Is your feature request related to a problem? Please describe

_No response_

### Describe your feature request!

I think the app should have a feature that when the line is drawn it automatically becomes straight if you stay still for a certain time with the pen pressed. It would be faster than using the ruler, which is still a useful tool. I like how it is implemented in onenote. I could try to take care of this. I would like to offer my contribution.

### Additional context

It is very convenient for example if you draw electrical circuits or anything else where you want to have a fast but not too precise workflow

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by MrGumpIT**
close #578

### CODE REVIEW

File: api/lib/src/models/tool.dart [hunk 1]
```diff
@@ -80,6 +80,7 @@ sealed class Tool with _$Tool {
     @Default('') String displayIcon,
     @Default(LabelMode.text) LabelMode mode,
     @Default(true) bool zoomDependent,
+    @Default(true) bool straightLineEnabled,
```
Comment (CodeDoctorDE) [inline api/lib/src/models/tool.dart:83]: please remove this here, label doesn't need it

File: app/lib/selections/tools/pen.dart [hunk 2]
```diff
@@ -21,6 +21,29 @@ class PenToolSelection extends ToolSelection<PenTool> {
                       zoomDependent: value ?? selected.first.zoomDependent))
                   .toList())),
       const SizedBox(height: 16),
+      // ! new checkBox
```
Comment (CodeDoctorDE) [inline app/lib/selections/tools/pen.dart:24]: I think this comment isn't necessary here

File: app/lib/selections/tools/pen.dart [hunk 3]
```diff
@@ -21,6 +21,29 @@ class PenToolSelection extends ToolSelection<PenTool> {
...
+      CheckboxListTile(
+          value: selected.first.straightLineEnabled,
+          title: Text(AppLocalizations.of(context).straightLineEnabled),
+          onChanged: (value) => update(
+              context,
+              selected
+                  .map((e) => e.copyWith(
+                      straightLineEnabled: value ?? selected.first.straightLineEnabled))
+                  .toList())),
+      // ! new slider
```
Comment (CodeDoctorDE) [inline app/lib/selections/tools/pen.dart:34]: here too and maybe we should hide the slider if the checkbox is unchecked

**Review by CodeDoctorDE — CHANGES_REQUESTED** [review]
Thanks for contributing. I have some things that I think we need to change

File: app/lib/selections/tools/pen.dart (hunk 2 above)
Comment (MrGumpIT) [reply app/lib/selections/tools/pen.dart:24]: done

File: api/lib/src/models/tool.dart (hunk 1 above)
Comment (MrGumpIT) [reply api/lib/src/models/tool.dart:83]: done

File: app/lib/selections/tools/pen.dart (hunk 3 above)
Comment (MrGumpIT) [reply app/lib/selections/tools/pen.dart:34]: done! Do you think it can go?

File: app/lib/l10n/app_en.arb [hunk 4]
```diff
@@ -151,6 +151,7 @@
   "copy": "Copy",
   "back": "Back",
   "zoomDependent": "Zoom dependent",
+  "straightLine": "Straight Line",
```
Comment (CodeDoctorDE) [inline app/lib/l10n/app_en.arb:154]:
```suggestion
  "shapeDetection": "Shape detection",
```

File: api/lib/src/models/tool.dart [hunk 5]
```diff
@@ -89,6 +89,8 @@ sealed class Tool with _$Tool {
     @Default('') String name,
     @Default('') String displayIcon,
     @Default(true) bool zoomDependent,
+    @Default(0.5) double straightLineTime,
+    @Default(false) bool straightLineEnabled,
```
Comment (CodeDoctorDE) [inline api/lib/src/models/tool.dart:93]:
```suggestion
    @Default(false) bool shapeDetectionEnabled,
```

File: api/lib/src/models/tool.dart [hunk 6]
```diff
@@ -89,6 +89,8 @@ sealed class Tool with _$Tool {
     @Default('') String name,
     @Default('') String displayIcon,
     @Default(true) bool zoomDependent,
+    @Default(0.5) double straightLineTime,
```
Comment (CodeDoctorDE) [inline api/lib/src/models/tool.dart:92]:
```suggestion
    @Default(0.5) double shapeDetectionTime,
```

**Review by CodeDoctorDE — COMMENTED** [review]
I would rename it here too to add more functionality in the future

**Comment by CodeDoctorDE** [comment]
I tested it out but don't get it working...
image
Also we should write the first character uppercase

File: app/lib/l10n/app_en.arb [hunk 7]
```diff
@@ -151,6 +151,8 @@
   "copy": "Copy",
   "back": "Back",
   "zoomDependent": "Zoom dependent",
+  "shapeDetection": "shape Detection",
```
Comment (CodeDoctorDE) [inline app/lib/l10n/app_en.arb:154]:
```suggestion
  "shapeDetection": "Shape Detection",
```

**Review by CodeDoctorDE — CHANGES_REQUESTED** [review]
I found one problem but then I can merge it

**Review by CodeDoctorDE — APPROVED** [review]
Thanks! LGTM

**Comment by HusniMuhammad** [comment]
> Thanks! LGTM

How it's works?

**Comment by CodeDoctorDE** [comment]
Draw a line, wait a bit, and then it should be converted to a line. You need to enable this feature in the pen tool properties view

**Pull request merged**